### DISCUSSION TITLE
ChatLocalAI (GPT4ALL) Request failed with status code 500

### DISCUSSION THREAD

**Discussion by hannahnguyen14** [FlowiseAI/Flowise#528] [Q&A]
I am trying to use ChatLocalAI as the chat (language) model but it keeps returning the error of "Request failed with status code 500".

I don't know what is the issue here because I have followed the ChatLocalAI setup tutorial (https://docs.flowiseai.com/chat-models/chatlocalai). I have done until the Test API part before setup chat flows by running this command "curl http://localhost:8080/v1/models" and it returns successfully like the guide. I don't know what to expect when open http://localhost:8080 on google chrome tab, I have tried to and it returns "{"error":{"code":404,"message":"Cannot GET /","type":""}}".

Moreover, I have tried ChatLocalAI with 3 different chat flows like below and they return the same error. The chat flow with OpenAI embedding and pinecone vector database works fine with other chat models like OpenAI and Azure as I have tried. 

Also, I have tried to change the base path to http://localhost:8080/ or add v1/models and it returns "failed with status code 404" so these paths seem wrong. Changing it to http://localhost:8080 base path returns the same error of 500.  

I have tried my best, please help me out! 
<img width="832" alt="Screenshot 2023-07-12 at 1 14 24 AM" src="https://github.com/FlowiseAI/Flowise/assets/112486498/fc5bcd3e-ee31-4f79-a440-469fdd586ea8">

<img width="1177" alt="Screenshot 2023-07-12 at 1 09 35 AM" src="https://github.com/FlowiseAI/Flowise/assets/112486498/287af082-3455-45a9-94b5-fbe8acd4a26c">

<img width="965" alt="Screenshot 2023-07-12 at 1 06 51 AM" src="https://github.com/FlowiseAI/Flowise/assets/112486498/906b37d2-d590-411f-99f2-b7c9217c0072">

**Comment by chungyau97** [comment]
Hi @hannahnguyen14,

Is your Flowise cloud hosted or run locally?

**Reply by hannahnguyen14** [reply]
Hi @chungyau97, I downloaded Flowise by following this guide so I think it is hosted locally: https://github.com/FlowiseAI/Flowise. I also downloaded LocalAI (https://docs.flowiseai.com/chat-models/chatlocalai) so it run locally too.

**Reply by chungyau97** [reply]
Hi @hannahnguyen14,

Sorry for the late reply.

Are you running [quick start](https://github.com/FlowiseAI/Flowise#quick-start), [docker](https://github.com/FlowiseAI/Flowise#-docker) or [setup](https://github.com/FlowiseAI/Flowise#setup)?

**Reply by hannahnguyen14** [reply]
Hi @chungyau97 , I'm running using docker

**Reply by chungyau97** [reply]
Hi @hannahnguyen14,

Try following this as a [reference example](https://docs.flowiseai.com/vector-stores/chroma#additional).

[LocalAI Port 8000](https://github.com/go-skynet/LocalAI/blob/master/docker-compose.yaml)

**Reply by hannahnguyen14** [reply]
Hi @chungyau97, thank you for your help. I have tried to follow to example and use http://host.docker.internal:8000 instead for the base path. However, it returns this error which I don't know how to fix either: getaddrinfo ENOTFOUND host.docker.internal.

Hope you can help me out!

**Reply by chungyau97** [reply]
My bad, I meant to type `8080`.

1. Modify `LocalAI\docker-compose.yaml`
```
version: '3.6'

networks:
  net:
    driver: bridge

services:
  api:
    image: quay.io/go-skynet/local-ai:latest
    build:
      context: .
      dockerfile: Dockerfile
    ports:
      - 8080:8080
    env_file:
      - .env
    volumes:
      - ./models:/models:cached
    networks:
      - net
    command: [ "/usr/bin/local-ai" ]
```
2. Build `LocalAI` and test following [ChatLocalAI](https://docs.flowiseai.com/chat-models/chatlocalai) and [LocalAI Embeddings](https://docs.flowiseai.com/embeddings/localai-embeddings)
3. Modify `Flowise\docker\docker-compose.yml`
```
version: '3.1'

services:
    flowise:
        image: flowiseai/flowise
        restart: always
        environment:
            - PORT=${PORT}
            - FLOWISE_USERNAME=${FLOWISE_USERNAME}
            - FLOWISE_PASSWORD=${FLOWISE_PASSWORD}
            - DEBUG=${DEBUG}
            - DATABASE_PATH=${DATABASE_PATH}
            - APIKEY_PATH=${APIKEY_PATH}
            - LOG_PATH=${LOG_PATH}
            - LOG_LEVEL=${LOG_LEVEL}
            - EXECUTION_MODE=${EXECUTION_MODE}
        ports:
            - '${PORT}:${PORT}'
        volumes:
            - ~/.flowise:/root/.flowise
        networks:
            - flowise_net
        command: /bin/sh -c "sleep 3; flowise start"

networks:
    flowise_net:
        name: localai_net
        external: true
```
4. Build Flowise following [Docker Compose](https://github.com/FlowiseAI/Flowise#docker-compose)

**Reply by hannahnguyen14** [reply]
Hi, I'm very sorry as I still can not figure it out. I really appreciate your help until now! So I have modified the files, built Local AI using docker-compose up -d --pull always, and I have built flowise beforehand. The error is still a 500 status code. I'm not using LocalAI embeddings as I don't really need it. Do you have any other suggestions for me? 

Again, thank you very much for taking the time to help me!

**Reply by chungyau97** [reply]
The LocalAI URL: `http://host.docker.internal:8080/v1`
Pinecone Dimensions: `384`
![image](https://github.com/FlowiseAI/Flowise/assets/33013947/4d187b9a-c08b-44ec-bc85-c5b4ba4df26d)

**Reply by hannahnguyen14** [reply]
Hi, I got trouble with base path http://host.docker.internal:8080/v1 as it returns this error: getaddrinfo ENOTFOUND host.docker.internal. Also, my pinecone dimension is 1536 instead, which I don't know if it cause any problems.

**Reply by chungyau97** [reply]
Troubleshoot steps:
1. Check `Flowise` and `LocalAI` use my scripts above to spin up; As my scripts above allow connection between 2 containers.
`docker-compose up -d`
2. Both `Flowise` and `LocalAI` are running on the docker container
**Windows Docker UI**:
![image](https://github.com/FlowiseAI/Flowise/assets/33013947/bd03ed11-843e-4075-bbd7-79311701ef5b)
**Linux**: `docker ps` or `sudo docker ps`
![image](https://github.com/FlowiseAI/Flowise/assets/33013947/464f9e1d-4f73-4857-bf7f-9a2e402ff866)
3. Check `LocalAI`'s models exists in Docker
![image](https://github.com/FlowiseAI/Flowise/assets/33013947/5d09805a-a670-4d41-b836-aa66d2af6a86)
4. `Pinecone`'s dimensions need to be 384. If you run into errors using Pinecone, please upgrade your Pinecone to a paid version.

**Reply by hannahnguyen14** [reply]
Oh, I see what's the problem here now. Sorry for the confusion, I set up Flowise running quick start before but I tried to set up LocalAI using docker. Is there any way I can run LocalAI using 'npx' too? 

Sorry for wasting your time, but I'm very positive about the process now as I can understand the problems underneath!

**Reply by HenryHengZJ** [reply]
might be related to this - https://github.com/FlowiseAI/Flowise/discussions/607

**Comment by matsuobasho** [comment]
> My bad, I meant to type `8080`.
> 
>     1. Modify `LocalAI\docker-compose.yaml`
> 
> 
> ```
> version: '3.6'
> 
> networks:
>   net:
>     driver: bridge
> 
> services:
>   api:
>     image: quay.io/go-skynet/local-ai:latest
>     build:
>       context: .
>       dockerfile: Dockerfile
>     ports:
>       - 8080:8080
>     env_file:
>       - .env
>     volumes:
>       - ./models:/models:cached
>     networks:
>       - net
>     command: [ "/usr/bin/local-ai" ]
> ```
> 
>     2. Build `LocalAI` and test following [ChatLocalAI](https://docs.flowiseai.com/chat-models/chatlocalai) and [LocalAI Embeddings](https://docs.flowiseai.com/embeddings/localai-embeddings)
> 
>     3. Modify `Flowise\docker\docker-compose.yml`
> 
> 
> ```
> version: '3.1'
> 
> services:
>     flowise:
>         image: flowiseai/flowise
>         restart: always
>         environment:
>             - PORT=${PORT}
>             - FLOWISE_USERNAME=${FLOWISE_USERNAME}
>             - FLOWISE_PASSWORD=${FLOWISE_PASSWORD}
>             - DEBUG=${DEBUG}
>             - DATABASE_PATH=${DATABASE_PATH}
>             - APIKEY_PATH=${APIKEY_PATH}
>             - LOG_PATH=${LOG_PATH}
>             - LOG_LEVEL=${LOG_LEVEL}
>             - EXECUTION_MODE=${EXECUTION_MODE}
>         ports:
>             - '${PORT}:${PORT}'
>         volumes:
>             - ~/.flowise:/root/.flowise
>         networks:
>             - flowise_net
>         command: /bin/sh -c "sleep 3; flowise start"
> 
> networks:
>     flowise_net:
>         name: localai_net
>         external: true
> ```
> 
>     4. Build Flowise following [Docker Compose](https://github.com/FlowiseAI/Flowise#docker-compose)

Does step 3 mean that I need to install flowise using Git and/or the Docker Compose method?  Like the OP, I just installed using the quick start and so I only have the LocalAI instance in Docker, but not the Flowise instance.  According to this discussion, it seems that to run LocalAi open models, I need to go down this route.

Additionally, if I need to use the Docker Compose method, then does that come after I perform the Git operations so that I can then add .env file?

**Reply by chungyau97** [reply]
Hi @matsuobasho,

Step 3 mean you will need to git clone Flowise repo and go through Docker Compose method. Reference: [Docker Compose README.md](https://github.com/FlowiseAI/Flowise/blob/main/docker/README.md)

If you're using these methods [[Quick Start](https://github.com/FlowiseAI/Flowise#quick-start), [Developers](https://github.com/FlowiseAI/Flowise#-developers)], you do not have to follow steps shown above.

**Reply by matsuobasho** [reply]
Just to be clear, in order to run ChatLocalAI do I need to have both LocalAI AND Flowise running in the Docker container?  From your guidance above, that seems to be the case.  I installed using the Quick Start method and only have LocalAI instance in Docker, not Flowise. 

So to go through the Docker Compose method, do I need to have the referenced file-structure, meaning that I need to install using the Developers approach (cloning the repo)?
![image](https://github.com/FlowiseAI/Flowise/assets/13874772/dc6a9a74-4e00-4195-a840-802e1c76e229)

Currently, I don't know which directory to add the .env file to.  Here are the contents of my Flowise directory in Windows under C/Users/Uname/AppData/Roaming/npm/node_modules/flowise/
![image](https://github.com/FlowiseAI/Flowise/assets/13874772/24aea7b5-f37c-4c65-85dc-b193985a2a4f)

Thanks!

**Reply by chungyau97** [reply]
> Hi @matsuobasho,
> 
> Step 3 mean you will need to git clone Flowise repo and go through Docker Compose method. Reference: [Docker Compose README.md](https://github.com/FlowiseAI/Flowise/blob/main/docker/README.md)
> 
> If you're using these methods [[Quick Start](https://github.com/FlowiseAI/Flowise#quick-start), [Developers](https://github.com/FlowiseAI/Flowise#-developers)], you do not have to follow steps shown above.

Hey @matsuobasho,

` If you're using these methods [Quick Start, Developers], you do not have to follow steps shown above.`
This mean you do not need to have Flowise on Docker to run ChatLocalAI

In Flowise, we have a total of 4 installation methods [Quick Start, Developer, Docker Compose, Docker Image]. Each method required different methods of connecting to ChatLocalAI
1. [Quick Start](https://github.com/flowiseai/flowise#quick-start) & [Developer](https://github.com/flowiseai/flowise#-developers) - Do not need to make Flowise into Docker in order to use ChatLocalAI
2. [Docker Compose](https://github.com/flowiseai/flowise#docker-compose) & [Docker Image](https://github.com/flowiseai/flowise#docker-image) - Require modification of Flowise file tin order to use ChatLocalAI

**Comment by matsuobasho** [comment]
Thanks for the detailed explanation.
So is the following expected behavior then?  I start LocalAI in Docker and then use the Quick Start method to launch Flowise.  

I test the API, the following returns info with StatusCode 200:
`curl http://localhost:8080/v1/models`

I build a simple chatbot in Flowise and get a `Request failed with status code 404` error when I ask the following question.  Note I double-checked the spelling of the model name.

![image](https://github.com/FlowiseAI/Flowise/assets/13874772/78045b2c-6240-4386-a6e5-d40079fb6888)

If I remove the http from the Base Path url, I get the following error:
`connect ENETUNREACH 0.0.31.144:80`

**Reply by chungyau97** [reply]
Hey @matsuobasho,

You got to use `http://host.docker.internal:8080/v1`

**Comment by matsuobasho** [comment]
Ok, thank you.  Two things:
1. I think that link should be in the [ChatLocalAI section](https://docs.flowiseai.com/chat-models/chatlocalai) of the tutorial (unless there's a way to get it to work with the original link http://localhost:8080/v1).  Would have probably saved me more than a few hours.  Happy to create a separate issue for that.
![image](https://github.com/FlowiseAI/Flowise/assets/13874772/2848456a-f93b-451b-89c7-189e40fe751c)

2. Why does it take such a long time to get a response?  First question took 1.5 minutes and a follow-up question is still processing 17 minutes after.  I realize this it outside the scope of this issue but grateful for any explanation or links!  Is it just querying the LocalAI LMs through Docker takes a long time?

**Reply by chungyau97** [reply]
Hi @matsuobasho 

1. It would be awesome if you could help us modified the docs.
- https://github.com/FlowiseAI/FlowiseDocs/blob/main/embeddings/localai-embeddings.md
- https://github.com/FlowiseAI/FlowiseDocs/blob/main/chat-models/chatlocalai.md

2. https://github.com/go-skynet/LocalAI/issues/646

**Reply by matsuobasho** [reply]
@chungyau97 , apologies for delay in response.  

1. Before I work on updating the docs with the correct path, please help me understand what I should update to.  i.e., is the original base path (`http://localhost:8080/v1`) valid for some scenarios of running LocalAI/ Flowise?  Or should I just replace it completely with the `http://api:8080` path? 
2. I have another workflow for asking questions about a document I load.  I'm running both Flowise and LocalAI through Docker.  As you can see below, I've specified the path mentioned in point 1.  I'm using the `text-embedding-ada-002.yaml` embeddings that are in the same LocalAI models directory.  And I get the same - `Error request failed with status code 500`.
![image](https://github.com/FlowiseAI/Flowise/assets/13874772/3e57552e-bf2f-48f2-891f-74c416a040f8)

As I mentioned [here](https://github.com/FlowiseAI/Flowise/discussions/528#discussioncomment-7007852), using the `api:8080` path does work with the simple ChatLocalAI + Prompt Template + LLM chain.

Thanks!

**Reply by chungyau97** [reply]
Hi @matsuobasho, it should be in `http://host.docker.internal:8080/v1`.
